password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,243 @@
+-- return true if given password is valid
+--
+-- This is equivalent to @null $ validatePassword policy password@
+--
+-- >>> let pass = mkPassword "This_Is_Valid_PassWord1234"
+-- >>> valid defaultPasswordPolicy pass
+-- True
+valid :: PasswordPolicy -> Password -> Bool
+valid policy pass = null $ validatePassword policy pass
+
+-- | Check if given 'Password' fulfills all of the Policies, returns list of
+-- reasons why it's invalid.
+--
+-- >>> let pass = mkPassword "This_Is_Valid_Password1234"
+-- >>> validatePassword defaultPasswordPolicy pass
+-- []
+validatePassword :: PasswordPolicy -> Password -> [InvalidReason]
+validatePassword passwordPolicy (Password password) =
+  catMaybes
+    [ isValidLength,
+      isTooShort,
+      isTooLong,
+      isUsingPolicyCharSet,
+      hasRequiredChar (passPolicyCharUppercase passwordPolicy) Uppercase,
+      hasRequiredChar (passPolicyCharLowercase passwordPolicy) Lowercase,
+      hasRequiredChar (passPolicyCharSpecial passwordPolicy) Special,
+      hasRequiredChar (passPolicyCharDigit passwordPolicy) Digit
+    ]
+  where
+    isValidLength :: Maybe InvalidReason
+    isValidLength =
+      if isValidPasswordPolicy passwordPolicy
+        then Nothing
+        else Just $ InvalidPasswordPolicy passwordPolicy
+    isTooLong :: Maybe InvalidReason
+    isTooLong =
+      let requiredMaxLength = passPolicyMaximumLength passwordPolicy
+      in if T.length password <= requiredMaxLength
+        then Nothing
+        else Just $ PasswordTooLong requiredMaxLength (T.length password)
+    isTooShort :: Maybe InvalidReason
+    isTooShort =
+      let requiredMinLength = passPolicyMinimumLength passwordPolicy
+       in if T.length password >= requiredMinLength
+            then Nothing
+            else Just $ PasswordTooShort requiredMinLength (T.length password)
+    isUsingPolicyCharSet :: Maybe InvalidReason
+    isUsingPolicyCharSet =
+        let filteredText = T.filter (\c -> c `notElem` (T.unpack $ passPolicyCharSet passwordPolicy)) password
+        in if T.null filteredText
+          then Nothing
+          else Just $ InvalidChar filteredText
+    hasRequiredChar :: Maybe Int -> CharacterCategory -> Maybe InvalidReason
+    hasRequiredChar Nothing _ = Nothing
+    hasRequiredChar (Just requiredCharNum) characterCategory =
+      let predicate = case characterCategory of
+            Uppercase -> isUpper
+            Lowercase -> isLower
+            Special   -> (\char -> char `elem` specialChars)
+            Digit     -> isDigit
+          actualRequiredCharNum = T.length $ T.filter predicate password
+       in if actualRequiredCharNum >= requiredCharNum
+            then Nothing
+            else Just $ NotEnoughReqChars characterCategory requiredCharNum actualRequiredCharNum
+
+-- | Checks if given 'PasswordPolicy' is valid
+--
+-- >>> isValidPasswordPolicy defaultPasswordPolicy
+-- True
+isValidPasswordPolicy :: PasswordPolicy -> Bool
+isValidPasswordPolicy policy =
```

Comment:
This code can be significantly reduced after shortening the fields and then using `RecordWildCards` here.
e.g. `isValidPasswordPolicy PasswordPolicy{..} =`